pkg/exemplar/exemplar.go

```diff
@@ -13,12 +13,86 @@
 
 package exemplar
 
-import "github.com/prometheus/prometheus/pkg/labels"
+import (
+	"bytes"
+	"encoding/json"
+	"math"
+	"strconv"
+
+	"github.com/prometheus/prometheus/pkg/labels"
+)
 
 // Exemplar is additional information associated with a time series.
 type Exemplar struct {
 	Labels labels.Labels
 	Value  float64
-	HasTs  bool
 	Ts     int64
+	HasTs  bool
+}
+
+// Equals compares if the exemplar e is the same as e2. Note that if HasTs is false for
+// both exemplars then the timestamps will be ignored for the comparison. This can come up
+// when an exemplar is exported without it's own timestamp, in which case the scrape timestamp
+// is assigned to the Ts field. However we still want to treat the same exemplar, scraped without/
+// an exported timestamp, as a duplicate of itself for each subsequent scrape.
+func (e Exemplar) Equals(e2 Exemplar) bool {
+	if e.Labels.String() != e2.Labels.String() {
+		return false
+	}
+
+	if (e.HasTs || e2.HasTs) && e.Ts != e2.Ts {
+		return false
+	}
+
+	if e.Value != e2.Value {
+		return false
+	}
+
+	return true
+}
+
+func (e Exemplar) MarshalJSON() ([]byte, error) {
+	var nts bytes.Buffer
+	partial := int(e.Ts / 1000)
+	fraction := int(e.Ts % 1000)
+
+	nts.Write([]byte(strconv.Itoa(partial)))
+	if fraction != 0 {
+		nts.WriteRune('.')
+		if fraction < 100 {
+			nts.WriteRune('0')
+
+		}
+		if fraction < 10 {
+			nts.WriteRune('0')
+		}
+		nts.Write([]byte(strconv.Itoa(fraction)))
+	}
+
+	f, err := strconv.ParseFloat(nts.String(), 64)
+	if err != nil {
+		return nil, err
+	}
```

Comment:
nit: can we have a comment about why we're serialising the exemplar are a string, re-parsing it, then serialising it to json again?

Reply:
Revamped the mashalling code to re-use parts from existing points.